Enable moveonly / noncopyable types by default

kavon opened the pull request:
This PR brings move-only types (still via `@_moveOnly`) out from behind the `-enable-experimental-move-only` flag. We still hold back features like `@_noImplicitCopy` and `_borrow`.

resolves rdar://106669967&106314333

kavon commented:
@swift-ci please test macos platform

kavon commented:
@swift-ci please smoke test

gottesmm commented on lib/SILGen/SILGenDecl.cpp:
```diff
@@ -468,8 +468,7 @@ class LetValueInitialization : public Initialization {
 public:
   LetValueInitialization(VarDecl *vd, SILGenFunction &SGF) : vd(vd) {
     const TypeLowering *lowering = nullptr;
-    if (SGF.getASTContext().LangOpts.Features.count(Feature::MoveOnly) &&
```
Please leave this in.

gottesmm commented on lib/SILGen/SILGenDecl.cpp:
```diff
@@ -509,8 +508,7 @@ class LetValueInitialization : public Initialization {
 
     // Make sure that we have a non-address only type when binding a
     // @_noImplicitCopy let.
-    if (SGF.getASTContext().LangOpts.Features.count(Feature::MoveOnly) &&
```
Same here.

gottesmm commented on lib/SILGen/SILGenDecl.cpp:
```diff
@@ -588,10 +586,6 @@ class LetValueInitialization : public Initialization {
     // If we have none...
     if (value->getOwnershipKind() == OwnershipKind::None) {
-      // If we don't have move only features enabled, just return, we are done.
-      if (!SGF.getASTContext().LangOpts.Features.count(Feature::MoveOnly))
```
This check is needed further down below since we handle no implicit copy ~609.

gottesmm commented on lib/SILGen/SILGenDecl.cpp:
```diff
@@ -1971,35 +1956,22 @@ void SILGenFunction::destroyLocalVariable(SILLocation silLoc, VarDecl *vd) {
         }
+      }
 
-        if (auto *cvi = dyn_cast<ExplicitCopyValueInst>(mvi->getOperand())) {
```
This probably needs to still be guarded by the feature. CopyableToMoveOnlyWrapper and MoveOnlyWrapperToCopyable are no implicit copy related.

gottesmm reviewed (commented):
Found some little issues

kavon committed:
run the move-checker if lexical borrow scopes are enabled

this pulls just the move checking passes out from
behind the experimental flag.

kavon commented:
@swift-ci please smoke test

kavon committed:
[bug fix] add missing lowering of explicit_copy_addr

kavon commented:
@swift-ci smoke test

kavon committed:
expose more of SILGen to move-only types by default

Also removes some redundant checks in SILGen for
the flag in order to process `_move` and `_borrow`
that are already checked for in Sema. We will keep
those behind the feature flag for now.

kavon committed:
move noncopyable types out from behind the feature flag

Since these types depend on lexical borrow scopes (a SIL
option), we're now making that the requirement whenever
such types appear.

kavon commented:
@swift-ci please test

kavon commented:
@swift-ci smoke test

kavon committed:
[bug fix] avoid generating explicit_copy_* for copyable types

The move checker was converting some kinds of copies
into their `explicit_copy_*` versions, despite the
type of the copy being a copyable type. This was
causing random crashes in some narrow circumstances.

resolves rdar://106669967

kavon committed:
basic verifier coverage for `explicit_copy_*` instructions

These are just basic restrictions for now to help avoid
mistakes. The restrictions are inferred based on how the
move checker uses these instructions.

kavon committed:
enable lexical borrow scopes for a test that needs it

kavon commented:
@swift-ci please test

kavon commented:
@swift-ci Please Test Source Compatibility

kavon committed:
update tests given move-only types are enabled

the main things still left behind the experimental flag(s) are
- move-only classes (guarded by MoveOnlyClasses feature)
- noimplicitcopy
- the _borrow operator

kavon committed:
enable full optimizations in test

doesn't seem to be a need to skip destroy hoisting
anymore.

kavon commented:
@swift-ci please test

kavon commented:
@swift-ci please test

kavon commented:
@swift-ci Please Test Source Compatibility

ktoso commented:
Wohoo, almost there! 🎉 🎉 🎉

kavon commented:
Debug source compat failure is with `swift-cluster-membership` and doesn't appear to be specific to this PR.
